packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts

```diff
@@ -0,0 +1,836 @@
+});
+      `,
+      errors: [
+        {
+          messageId: 'rejectAnError',
+          type: AST_NODE_TYPES.CallExpression,
+          line: 3,
+          endLine: 3,
+          column: 3,
+          endColumn: 12,
+        },
+      ],
+    },
+    {
+      code: `
+new Promise(function (foo, arguments) {
+  arguments(5);
+});
+      `,
+      errors: [
+        {
+          messageId: 'rejectAnError',
+          type: AST_NODE_TYPES.CallExpression,
+          line: 3,
+          endLine: 3,
+          column: 3,
+          endColumn: 15,
+        },
+      ],
+    },
+    {
+      code: 'new Promise((foo, arguments) => arguments(5));',
+      errors: [
+        {
+          messageId: 'rejectAnError',
+          type: AST_NODE_TYPES.CallExpression,
+          line: 1,
+          endLine: 1,
+          column: 33,
+          endColumn: 45,
+        },
+      ],
+    },
+    {
+      code: `
+new Promise(function ({}, reject) {
+  reject(5);
+});
+      `,
+      errors: [
+        {
+          messageId: 'rejectAnError',
+          type: AST_NODE_TYPES.CallExpression,
+          line: 3,
+          endLine: 3,
+          column: 3,
+          endColumn: 12,
+        },
+      ],
+    },
+    {
+      code: 'new Promise(({}, reject) => reject(5));',
+      errors: [
+        {
+          messageId: 'rejectAnError',
+          type: AST_NODE_TYPES.CallExpression,
+          line: 1,
+          endLine: 1,
+          column: 29,
+          endColumn: 38,
+        },
+      ],
+    },
+    {
+      code: 'new Promise((resolve, reject, somethingElse = reject(5)) => {});',
+      errors: [
+        {
+          messageId: 'rejectAnError',
+          type: AST_NODE_TYPES.CallExpression,
+          line: 1,
+          endLine: 1,
+          column: 47,
+          endColumn: 56,
+        },
+      ],
+    },
+    {
+      code: `
+declare const foo: {
+  bar: PromiseConstructor;
+};
+new foo.bar((resolve, reject) => reject(5));
+      `,
+      errors: [
+        {
+          messageId: 'rejectAnError',
+          type: AST_NODE_TYPES.CallExpression,
+          line: 5,
+          endLine: 5,
+          column: 34,
+          endColumn: 43,
+        },
+      ],
+    },
+    {
+      code: `
+declare const foo: {
+  bar: PromiseConstructor;
+};
+new (foo?.bar)((resolve, reject) => reject(5));
+      `,
+      errors: [
+        {
+          messageId: 'rejectAnError',
+          type: AST_NODE_TYPES.CallExpression,
+          line: 5,
+          endLine: 5,
+          column: 37,
+          endColumn: 46,
+        },
+      ],
+    },
+    {
+      code: `
+const foo = Promise;
+new foo((resolve, reject) => reject(5));
+      `,
+      errors: [
+        {
+          messageId: 'rejectAnError',
+          type: AST_NODE_TYPES.CallExpression,
+          line: 3,
+          endLine: 3,
+          column: 30,
+          endColumn: 39,
+        },
+      ],
+    },
+    {
+      code: `
+class Foo extends Promise<number> {}
+Foo.reject(5);
+      `,
```

Comment:
Thanks to @Tjstretchalot for pointing on this! https://github.com/typescript-eslint/typescript-eslint/pull/8094#issuecomment-1864771039